프로젝트 관리 앱 [STEP 2-2] 아샌, 앨리

ICS-Asan opened the pull request:
안녕하세요. 찰리! 👾 @kcharliek
아샌(@ICS-Asan), 앨리(@Allie-e)입니다. 

지난 번 말씀 드렸던 STEP2-2 구현이 완료되어 PR 올립니다! 

Rx사용과 MVVM패턴 적용이 어려웠던 만큼 부족한 부분도 많겠지만 잘 부탁드립니다!!🙏🏻

감사합니다!! 

# 💻  구현 내용

### <  ProjectViewModel >

> MVVM 패턴을 적용하기 위해 ProjectViewModel을 생성해 주었습니다.
지난번 보내주신 [링크](https://github.com/sergdort/CleanArchitectureRxSwift)와  추가로 설명해주신 것을 토대로 구현을 해봤습니다.
저희가 View와 ViewModel사이에 주고 받아야 할 시점을 고려하여 아래와 같이Observable타입들을 구현하였습니다.
1. Move To ToDo, Doing, Done을 눌렀을 때
2. cell이 선택되었을 때
3. 삭제할 때
4. 데이터 처리가 끝나고 view의 reload가 필요할 때
최대한 ViewModel에서 데이터 처리를 할 수 있도록 하기 위해 View에서는 해당 시점에서 어떤 프로젝트인지 알려주는 역할만 할 수 있게 고민하여 구현했습니다.
그리고 데이터를 담아 둘 `projects`를 활용하여 데이터 처리기능을 구현했습니다.
현재는 데이터를 따로 저장하고 있지만 추후에 해당 기능이 들어가면 모델에서 불러온 데이터가 담아질 수 있을 것 같습니다.
데이터를 처리하는 과정에서 View에서 작업이 필요한 project를 받아 ViewModel이 가지고 있는 projects에서 같은 요소를 찾은 뒤 필요한 작업을 처리하고 reload할 수 있도록 했습니다.
> 

### <  ProjectListCellDelegate >

> 기존 각 TableView에 있던 LongPressGesture를 Cell로 옮겼습니다.
그 과정에서 popover로 띄운 메뉴를 탭했을 때 해야 할 작업을 구현해야하는데 Cell내부에서 구현할 수가 없었습니다.
ViewModel을 따로 만들어서 관리하는 것은 현재 단계에서 어떻게 구현해야할지 떠오르지가 않아 Delegate를 사용하기로 했습니다.
Delegate를 생성하여 MainViewController에 채택해주었고 Delegate메서드 내부에 ViewModel에 작업을 지시하도록 구현했습니다.
> 

```swift
extension MainViewController: ProjectListCellDelegate {
    func didTapTodoAction(_ project: Project?) {
        guard let project = project else { return }
        moveToToDoObserver.onNext(project)
    }
    
    func didTapDoingAction(_ project: Project?) {
        guard let project = project else { return }
        moveToDoingObserver.onNext(project)
    }
    
    func didTapDoneAction(_ project: Project?) {
        guard let project = project else { return }
        moveToDoneObserver.onNext(project)
    }
    
    func presentPopover(_ alert: UIAlertController) {
        present(alert, animated: true)
    }
}
```

### <  Project >

> 데이터를 관리하기 위해 Project에 관련된 타입이 필요하다고 생각되어 해당 타입을 구현했습니다.
그리고 formattedDate, isExpired와 같이 Project를 사용하면서 필요한 특정 값들을 사용할 수 있게 연산 프로퍼티로 구현했습니다.
> 

### <  DateFormatter Extension >

> DateFormatter의 extension에 싱글턴으로 DateFormatter로 구현하여 필요한 세팅을 해주었습니다.
DateFormatter의 경우 특히 cell을 만드는 과정에서 자주 호출 될 것으로 판단해 필요할 때마다 계속 DateFormatter를 생성하는 것 보다는 싱글턴으로 구현해놓는 것이 더 효율적이라고 생각했습니다.
> 

## 🔫 Trouble Shooting

### < 프로젝트 기한 만료시 `textColor = .systemRed`가 작동하지 않는 문제 >

`원인` 아래와같이 `setupDateLabel(with:)` 메서드로 프로젝트 기한이 만료되면 `dateLabel.textColor = .systemRed`로 표기되도록 조건문을 걸어주었으나 제대로 작동하지 않았다.
![image](https://user-images.githubusercontent.com/91048487/158763627-8906a4fc-bba2-4ac8-b235-3051a71ee916.png)

`setSelected()` 메서드에서 else문을 타면서 셀이 선택되지 않은 상태에서는 `dateLabel.textColor = .label`로 지정되는 것이 문제였다.

![image](https://user-images.githubusercontent.com/91048487/158763647-b1a63095-3463-49b3-997e-bf2d9bdcac51.png)

`해결` `setSelected()` 메서드 내부의 dateLabel을 지정해주는 부분에 `setupDateLabel(with:)` 메서드를 호출해서 프로젝트 날짜를 한번 검사하고 설정해주도록 변경해주었다.

![image](https://user-images.githubusercontent.com/91048487/158763673-184f71a5-ecb5-4956-a79b-185bc16e9a60.png)

## ❓   질문할 내용

### < tableView() 메서드 내부 반복되는 코드를 피하는 방법 >

ViewModel를 생성하면서 tableViewDataSource를 각각 만들어 사용하는 것은 어렵다고 생각되어 하나의 DataSource에서 처리하는 방법으로 변경하였습니다.
이 과정에서 각각의 테이블 뷰마다 cell에 대한 세팅을 해주는 코드가 반복되고있어서 개선을 해주고 싶은데 좋은 방법이 떠오르지 않습니다,, 어떤 방식으로 접근하면 좋을까요? 🥲

![image](https://user-images.githubusercontent.com/91048487/158763702-2094b1af-378f-46cb-a8e4-eed2fca12d7c.png)

### < SPM 에러.....>

아샌이 push 하고 제가 pull 받으면 이런 에러가 뜹니다.. 아샌이 암살시도 한거겠죠?ㅎ

해결은,, 프로젝트 폴더에서 Package.resolved 파일을 삭제하고 xcode에서 package cache 도 삭제한 후 resolving packgae를 새로 해주는 방법으로 하긴 했는데 pull 받을 때마다 이런 문제가 반복됩니다..

왜 그런건지 혹시 아시나요ㅠㅠ?

![image](https://user-images.githubusercontent.com/91048487/158763726-11cd9789-db20-4d1f-b62b-a917ca441383.png)

감사합니다!!👍

Allie-e committed:
feat: Project 이동 기능 구현 #3

- Model 및 ViewModel 생성
- LongPressGesture cell에 등록
- ProjectListCellDelegate 구현 및 MainViewController로 지정

ICS-Asan committed:
feat: project 추가 기능 구현 #4

Allie-e committed:
feat: project edit 기능 구현 #4

ICS-Asan committed:
feat: project 삭제 기능 구현 #4

- 기존 selectedProject를 할당해주는 과정을 selectedObserver.onNext로 처리

Allie-e committed:
refactor: 기타 UI 작업 #5

- 프로젝트 날짜 비교해서 만료된 날짜는 빨간글씨로 표기
- EditView에서 해당 셀의 state가 navigation title에 반영되도록 함
- dateFormatter extension으로 싱글턴 생성하여 활용
- 각 section의 project 개수를 보여주는 badgeLabel 수정

ICS-Asan committed:
refactor: Project 내부 isExpired 생성 및 파일 배치 수정 #5

kcharliek commented on ProjectManager/ProjectManager/Controller/AddProjectViewController.swift:
```diff
@@ -49,8 +50,10 @@ class AddProjectViewController: UIViewController {
-}
+    
+    @objc private func addProject() {
+        viewModel?.addProject(title: addView.title, body: addView.body, date: addView.date)
```
'done button을 눌렀을 때 addProject를 한다' 라는 내용은 'Business Logic' 입니다.
Click Event를 ViewModel로 보내서 ViewModel이 판단하도록 수정해주세요.

Allie-e replied:
네 반영했습니다!

kcharliek commented on ProjectManager/ProjectManager/Controller/EditProjectViewController.swift:
```diff
@@ -2,13 +2,22 @@ import UIKit
 
 class EditProjectViewController: UIViewController {
     private let editView = ProjectFormView()
+    weak var viewModel: ProjectViewModel?
```
다른 View의 ViewModel을 직접 가져오면 안됩니다.
EditProjectViewModel을 추가하고, ProjectViewModel의 기능을 어떻게 가져오는게 좋을지. 다시 고민해주세요.

kcharliek commented on ProjectManager/ProjectManager/Controller/EditProjectViewController.swift:
```diff
@@ -47,9 +62,14 @@ class EditProjectViewController: UIViewController {
         actionAfterDismiss()
     }
+    
+    @objc private func saveEditedProject() {
```
Business Logic 입니다.

kcharliek commented on ProjectManager/ProjectManager/Controller/MainViewController.swift:
```diff
@@ -1,17 +1,22 @@
+    private let moveToToDoObserver: PublishSubject<Project> = .init()
+    private let moveToDoingObserver: PublishSubject<Project> = .init()
+    private let moveToDoneObserver: PublishSubject<Project> = .init()
+    private let selectObserver = PublishSubject<Project>.init()
```
코딩 스타일을 통일해주세요.

kcharliek commented on ProjectManager/ProjectManager/Controller/MainViewController.swift:
```diff
@@ -1,17 +1,22 @@
+import RxSwift
 
 class MainViewController: UIViewController {
+    private let viewModel = ProjectViewModel()
```
MainViewController 의 ViewModel이 ProjectViewModel 인 것이 어색하게 느껴집니다.

Allie-e replied:
MainViewModel로 변경하도록 하겠습니다!

kcharliek commented on ProjectManager/ProjectManager/Controller/MainViewController.swift:
```diff
@@ -113,85 +155,95 @@ class MainViewController: UIViewController {
+            return UITableViewCell()
+        }
+        switch tableView {
+        case toDoTableView:
```
코드 반복을 어떻게 하면 줄일 수 있을지 고민해주세요.

kcharliek replied:
코멘트 남기고 보니, 질문 주신 부분이었네요.

tableView 의 type에 따라 todoProjects/doingProjects/doneProjects 를 선택하면 되니까요.

`func projects(for tableView: UITableView) -> [Project]`

같은 메서드를 만들면 되지않을까요.

Allie-e replied:
말씀해주신대로 테이블뷰를 파라미터로 받아 프로젝트를 리턴하는 함수를 만들어 중복되는 코드 삭제해주었습니다!

kcharliek commented on ProjectManager/ProjectManager/Controller/MainViewController.swift:
```diff
@@ -113,85 +155,95 @@ class MainViewController: UIViewController {
+    }
+    
     func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard let cell = tableView.cellForRow(at: indexPath) as? ProjectListCell,
```
cell을 select했을 때 EditProjectViewController 을 띄우는 건 Business Logic입니다.

kcharliek commented on ProjectManager/ProjectManager/View/ProjectListCell.swift:
```diff
@@ -1,6 +1,9 @@
 import UIKit
 
 class ProjectListCell: UITableViewCell {
+    var delegate: ProjectListCellDelegate?
```
strong delegate ?

Allie-e replied:
weak로 수정해주었습니다~!

kcharliek commented on ProjectManager/ProjectManager/View/ProjectListCell.swift:
```diff
@@ -1,6 +1,9 @@
 
 class ProjectListCell: UITableViewCell {
+    var delegate: ProjectListCellDelegate?
+    var project: Project?
```
private 챙겨주세요 .

kcharliek commented:
## 질문 주신 내용에 대해

1. tableView 관련 코드 반복

코드에 코멘트 드렸습니다.

2. SPM 오류에 대해.

글쎄요.. 이런 문제는 겪어본 적이 없네요. 구글링 해보아야할 것 같습니다. 
gitignore 확인, Xcode 버전 서로 다른지 확인. 정도 권해드립니다.

Allie-e committed:
refactor: ProjectViewModel 네이밍 MainViewModel로 변경 #6

- tableView 메서드 내부 중복 코드 제거
- delegate weak로 변경

ICS-Asan committed:
refactor: 비즈니스 로직 ViewModel로 이동 #6

- add/edit viewModel 생성
- ProjectRepository 생성
- cell이 직접 데이터를 갖고 있지 않도록 로직 변경
- typealias 생성 및 활용

Allie-e committed:
refactor: AddProjectView, ViewModel Rx 적용 #6

ICS-Asan committed:
refactor: EditProjectView, ViewModel Rx 적용 #6

Allie-e committed:
refactor: MainViewController 내부 business logic ViewModel로 이동 #6

kcharliek reviewed (approved):
MVVM 그림이 나오네요 ㅎㅎ 아주 잘 해주셨습니다.
